Add support for embedded objects

hsul4n opened the pull request:
Here's an implementation for `@Embedded` annotation with test and example. 

Closing #9

hsul4n committed:
Add @Embedded annotation

hsul4n committed:
- Add embedded value_object and processor.
- Use prefix in field_processor for prefixing embedded columns.
- Use StringBuffer instead of List<String> for building the constructor on queryable_processor for more clarity .
- Add embedded_processor_test for testing embeddings.

hsul4n commented:
Is there any issue with testing nullable boolean

vitusortner commented:
Thanks for your contribution! I'll look into it as soon as possible.

hsul4n committed:
Implement embedded objects in example

hsul4n committed:
Update floor version to 0.13.0 & update README.md for implementing embedded objects

hsul4n commented:
@vitusortner sorry for non-nessory commit, check out the new clarified commits.

hsul4n commented:
Rebasing develop branch to get the latest commit #327

vitusortner commented on floor_generator/lib/processor/queryable_processor.dart:
```diff
@@ -21,33 +23,81 @@ abstract class QueryableProcessor<T extends Queryable> extends Processor<T> {
+          .join(', '),
+    );
+
+    for (final embedded in embeddeds) {
+      final parameterName = embedded.fieldElement.displayName;
+      final field = constructorParameters.firstWhere(
+        (parameterElement) => parameterElement.name == parameterName,
+        orElse: () => null,
+      );
+
+      constructorBuffer.write(', ');
+
+      if (field.isNamed) {
+        constructorBuffer.write('$parameterName: ');
+      }
+
+      constructorBuffer.write('${embedded.classElement.displayName}(');
+
+      constructorBuffer.write(
+        embedded.classElement.constructors.first.parameters
+            .map((parameterElement) =>
+                _getParameterValue(parameterElement, embedded.fields))
+            .where((parameterValue) => parameterValue != null)
+            .join(', '),
+      );
+
+      constructorBuffer.write(')');
+    }
```
❓Does this mean that embedded fields can only be the last constructor parameters?
The following wouldn't work, right?
```dart
class Person {
  @primaryKey
  final int id;

  @embedded
  final Address address;

  final String name;

  Person(this.id, this.address, this.name);
}

class Address {
  final String street;

  Address(this.street);
}
```

vitusortner commented on floor_generator/lib/processor/queryable_processor.dart:
```diff
@@ -103,6 +153,10 @@ abstract class QueryableProcessor<T extends Queryable> extends Processor<T> {
+    return !(isStatic || isSynthetic || isIgnored || shouldEmbedded());
+  }
+
+  bool shouldEmbedded() {
```
💡Naming idea: `shouldBeEmbedded()` or `isEmbedded` getter instead of function.

vitusortner commented on floor_generator/lib/processor/queryable_processor.dart:
```diff
@@ -21,33 +23,81 @@ abstract class QueryableProcessor<T extends Queryable> extends Processor<T> {
       : assert(classElement != null),
         _queryableProcessorError = QueryableProcessorError(classElement);
 
+  List<FieldElement> get _allFields => [
+        ...classElement.fields,
+        ...classElement.allSupertypes.expand((type) => type.element.fields),
+      ];
```
💡This can be moved to the constructor so it's only being initialized once instead of on every access.

vitusortner commented on floor_generator/lib/processor/embedded_processor.dart:
```diff
@@ -0,0 +1,46 @@
+  }
+
+  @nonNull
+  List<Field> _getFields() {
+    return _classElement.fields
+        .map((fieldElement) =>
+            FieldProcessor(fieldElement, _getPrefix()).process())
+        .toList();
+  }
```
❓Does this mean that nested embeddables aren't supported?
E.g.
```dart
class Person {
  @primaryKey
  final int id;

  @embedded
  final Address address;

  final String name;

  Person(this.id, this.address, this.name);
}

class Address {
  @embedded
  final Street street;

  Address(this.street);
}

class Street {
  final String street;

  Street(this.street);
}
```

vitusortner commented on floor_generator/lib/value_object/embedded.dart:
```diff
@@ -0,0 +1,28 @@
+  final FieldElement fieldElement;
+  final List<Field> fields;
+
+  ClassElement get classElement => fieldElement.type.element as ClassElement;
```
💡This can be moved to the constructor so it's only being initialized once instead of on every access.

vitusortner commented on floor_generator/lib/value_object/entity.dart:
```diff
@@ -1,6 +1,7 @@
 import 'package:analyzer/dart/element/element.dart';
 import 'package:collection/collection.dart';
 import 'package:floor_generator/misc/annotations.dart';
+import 'package:floor_generator/value_object/embedded.dart';
```
Please add tests for all the changes in this class.

vitusortner commented on floor_generator/lib/value_object/entity.dart:
```diff
@@ -30,6 +32,13 @@ class Entity extends Queryable {
       return field.getDatabaseDefinition(autoIncrement);
     }).toList();
 
+    final embeddedDefinitions = embeddeds.map((embed) {
```
💡 
```suggestion
    final embeddedDefinitions = embeddeds.map((embedded) {
```

vitusortner reviewed (commented):
I've left some questions. Thanks for your initiative in starting this non-trivial task!

hsul4n commented:
Thanks, @vitusortner for catching gaps, I'll try to follow the suggestions and also provide better testing.

hsul4n committed:
- Add support for nested embedded objects.
- Add boolean extension for parsing nullable fields.
- Add Fieldable abstract class for both (Field, Embedded).
- Passing null in constructor when field is ignore.
- Fix nullable boolean field test case.
- Some improvement and code enhancement.

hsul4n commented:
> I've left some questions. Thanks for your initiative in starting this non-trivial task!

Resolved with 292cfb425fa9ce2d4d07733df501c988fee24033, will push new `CHANGELOG.md` & version changes after approving changes.

proninyaroslav commented:
@hsul4n  Will this feature reach the final point?

hsul4n commented:
> @hsul4n  Will this feature reach the final point?

Absolutely, but for now theres a major change on api due to #318 & #361, also embedded objects depend on relations annotations to reach final points.

proninyaroslav commented:
@hsul4n 
Thank you. Yeah, I really look forward to the support of type converters, as well as the embedded objects. Without them, I will to make very ugly code.

vitusortner committed:
Merge branch 'develop' into feat_embedded

proninyaroslav commented:
@hsul4n 
During this time, a lot of work has been done, including the appearance of type converters. It would be great to have also embedded objects 👍

amkuchta commented:
@hsul4n @vitusortner I'm really looking forward to this feature - Floor is exactly what I was looking for in an ORM package, and this is the missing piece that I need.  Thank you guys for all of the hard work!

Is there a rough idea of when it will be completed?

themadmrj commented:
Any news on embedded & relation features?
ATM I need 3 subsequent queries with post processing to achieve what could be done in a single query :(

amkuchta commented:
@hsul4n @vitusortner any updates on this?  It's been about 6 months since I last asked, and it seems to have died... Looks like no changes since 30 July 2020?

marcosmko commented:
any updates?
maybe I could finish this, if needed!

amkuchta commented:
@marcosmko I'd be willing to pitch in, too. I'll need to poke through the code to get an idea if how it all works

vitusortner commented:
@marcosmko, @amkuchta anything I can help you with? This PR has a bunch of conflicts by now but it might still help to get an idea of how to implement the feature. Thanks for your support 🙏

marcosmko commented:
Hello @vitusortner, @amkuchta and @hsul4n, I synced this branch and created #538, but I still need to fix some things that got broken.

mobile-pablo commented:
Hi, whats status after all those years? I need embedded objects
